### ISSUE TITLE
Pyodide support for running in jupyter-lite browser-based notebooks

### ISSUE THREAD

**Issue by ryuwd** [scikit-hep/uproot5#854]
Carrying on discussion here: https://matrix.to/#/!WwIPZwMtNCFuzXqrmf:gitter.im/$XZVjnFnoDutfxGE9GcfzDUToky0yiQGCKVvYvKy9s90?via=gitter.im&via=matrix.org

It'd be really interesting to get `uproot` running in `jupyter-lite` under the pyodide kernel, so far I have the following recipe:

```py
import piplite

%pip install -q numpy pandas pyodide-http

import pyodide_http

pyodide_http.patch_all()
from http import client
class HTTPSConnection: pass
client.HTTPSConnection = HTTPSConnection

%pip install --no-deps uproot
```

However, one cannot get around that `HTTPSource` relies on ThreadPools, leading to this exception:

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[3], line 1
----> 1 f=uproot.open("https://github.com/scikit-hep/scikit-hep-testdata/raw/main/src/skhep_testdata/data/uproot-hepdata-example.root")

File /lib/python3.10/site-packages/uproot/reading.py:141, in open(path, object_cache, array_cache, custom_classes, decompression_executor, interpretation_executor, **options)
    132 if not uproot._util.isstr(file_path) and not (
    133     hasattr(file_path, "read") and hasattr(file_path, "seek")
    134 ):
    135     raise ValueError(
    136         "'path' must be a string, pathlib.Path, an object with 'read' and "
    137         "'seek' methods, or a length-1 dict of {file_path: object_path}, "
    138         f"not {path!r}"
    139     )
--> 141 file = ReadOnlyFile(
    142     file_path,
    143     object_cache=object_cache,
    144     array_cache=array_cache,
    145     custom_classes=custom_classes,
    146     decompression_executor=decompression_executor,
    147     interpretation_executor=interpretation_executor,
    148     **options,  # NOTE: a comma after **options breaks Python 2
    149 )
    151 if object_path is None:
    152     return file.root_directory

File /lib/python3.10/site-packages/uproot/reading.py:581, in ReadOnlyFile.__init__(self, file_path, object_cache, array_cache, custom_classes, decompression_executor, interpretation_executor, **options)
    576 self.hook_before_create_source()
    578 Source, file_path = uproot._util.file_path_to_source_class(
    579     file_path, self._options
    580 )
--> 581 self._source = Source(
    582     file_path, **self._options  # NOTE: a comma after **options breaks Python 2
    583 )
    585 self.hook_before_get_chunks()
    587 if self._options["begin_chunk_size"] < _file_header_fields_big.size:

File /lib/python3.10/site-packages/uproot/source/http.py:564, in HTTPSource.__init__(self, file_path, **options)
    562 self._fallback_options = options.copy()
    563 self._fallback_options["num_workers"] = self._num_fallback_workers
--> 564 self._open()

File /lib/python3.10/site-packages/uproot/source/http.py:567, in HTTPSource._open(self)
    566 def _open(self):
--> 567     self._executor = uproot.source.futures.ResourceThreadPoolExecutor(
    568         [HTTPResource(self._file_path, self._timeout)]
    569     )

File /lib/python3.10/site-packages/uproot/source/futures.py:351, in ResourceThreadPoolExecutor.__init__(self, resources)
    349     self._workers.append(ResourceWorker(self._work_queue, resource))
    350 for worker in self._workers:
--> 351     worker.start()

File /lib/python3.10/threading.py:928, in Thread.start(self)
    926     _limbo[self] = self
    927 try:
--> 928     _start_new_thread(self._bootstrap, ())
    929 except Exception:
    930     with _active_limbo_lock:

RuntimeError: can't start new thread
```

It seems like the next step would be to create a pyodide-friendly `HTTPSource` that uses either `pyodide-http` or `js.fetch` interfaces.

cc @chrisburr @agoose77

**Comment by agoose77** [comment]
This is really exciting! In principle, there's not a lot stopping us from using `uproot` in the browser. At this point, it probably makes sense for the upstream dependency (`awkward`) to be distributed on both emscripten-forge and pyodide. We already build pyodide-wheels, but we build the package ourselves. 

Uproot itself should be fairly trivial; we don't have any other _required_ binary dependencies. The assumption about various packages that do not exist in pyodide being available is fairly easy to resolve; we can add some guards for emscripten. The challenge will probably be in our optional dependencies used for compression. These are not _really_ optional; most ROOT files are compressed in one form or another. 

I can see uproot-in-browser as a useful thing; a big selling point of pyodide (for me) is that it's very easy to spin up, and is highly portable. If we start packaging uproot and awkward in the aforementioned ecosystems, we can provide a use case for packaging these other wheels (xxhash, lz4, etc).

**Comment by jpivarski** [comment]
It should work, and I'm all for it, except—how do you end up with ROOT files in the browser? They have to be explicitly copied in or something. Maybe the HTTP backend? They couldn't be very large...

**Comment by ryuwd** [comment]
> If the issue is threading, maybe switching things to TrivialExecutor would work.

Here's all the code I used in my test jupyterlite notebook:
```py
import piplite

%pip install -q numpy pandas pyodide-http

import pyodide_http

pyodide_http.patch_all()
from http import client
class HTTPSConnection: pass
client.HTTPSConnection = HTTPSConnection

%pip install --no-deps uproot

import uproot
uproot.decompression_executor = uproot.source.futures.TrivialExecutor
uproot.interpretation_executor = uproot.source.futures.TrivialExecutor


f=uproot.open("https://github.com/scikit-hep/scikit-hep-testdata/raw/main/src/skhep_testdata/data/uproot-hepdata-example.root")
```
and the traceback:
```
RuntimeError                              Traceback (most recent call last)
Cell In[3], line 1
----> 1 f=uproot.open("https://github.com/scikit-hep/scikit-hep-testdata/raw/main/src/skhep_testdata/data/uproot-hepdata-example.root")

File /lib/python3.10/site-packages/uproot/reading.py:141, in open(path, object_cache, array_cache, custom_classes, decompression_executor, interpretation_executor, **options)
    132 if not uproot._util.isstr(file_path) and not (
    133     hasattr(file_path, "read") and hasattr(file_path, "seek")
    134 ):
    135     raise ValueError(
    136         "'path' must be a string, pathlib.Path, an object with 'read' and "
    137         "'seek' methods, or a length-1 dict of {file_path: object_path}, "
    138         f"not {path!r}"
    139     )
--> 141 file = ReadOnlyFile(
    142     file_path,
    143     object_cache=object_cache,
    144     array_cache=array_cache,
    145     custom_classes=custom_classes,
    146     decompression_executor=decompression_executor,
    147     interpretation_executor=interpretation_executor,
    148     **options,  # NOTE: a comma after **options breaks Python 2
    149 )
    151 if object_path is None:
    152     return file.root_directory

File /lib/python3.10/site-packages/uproot/reading.py:581, in ReadOnlyFile.__init__(self, file_path, object_cache, array_cache, custom_classes, decompression_executor, interpretation_executor, **options)
    576 self.hook_before_create_source()
    578 Source, file_path = uproot._util.file_path_to_source_class(
    579     file_path, self._options
    580 )
--> 581 self._source = Source(
    582     file_path, **self._options  # NOTE: a comma after **options breaks Python 2
    583 )
    585 self.hook_before_get_chunks()
    587 if self._options["begin_chunk_size"] < _file_header_fields_big.size:

File /lib/python3.10/site-packages/uproot/source/http.py:564, in HTTPSource.__init__(self, file_path, **options)
    562 self._fallback_options = options.copy()
    563 self._fallback_options["num_workers"] = self._num_fallback_workers
--> 564 self._open()

File /lib/python3.10/site-packages/uproot/source/http.py:567, in HTTPSource._open(self)
    566 def _open(self):
--> 567     self._executor = uproot.source.futures.ResourceThreadPoolExecutor(
    568         [HTTPResource(self._file_path, self._timeout)]
    569     )

File /lib/python3.10/site-packages/uproot/source/futures.py:351, in ResourceThreadPoolExecutor.__init__(self, resources)
    349     self._workers.append(ResourceWorker(self._work_queue, resource))
    350 for worker in self._workers:
--> 351     worker.start()

File /lib/python3.10/threading.py:928, in Thread.start(self)
    926     _limbo[self] = self
    927 try:
--> 928     _start_new_thread(self._bootstrap, ())
    929 except Exception:
    930     with _active_limbo_lock:

RuntimeError: can't start new thread
```

The thread issue points to `HTTPSource` using `ThreadPoolExecutor`

> It should work, and I'm all for it, except—how do you end up with ROOT files in the browser? They have to be explicitly copied in or something. Maybe the HTTP backend? They couldn't be very large...

There is an interface to fetching content with the browser:
https://jupyterlite.readthedocs.io/en/0.1.0-beta/howto/content/python.html#fetching-remote-content

We wanted to see if we could give people a quick way to setup a notebook in their browser, serve the ROOT files over HTTP.

The more obvious solution would be to generate a notebook, have the user download and run it locally, but running in the browser would be interesting to try - if not just to see how feasible it is.

**Comment by chrisburr** [comment]
> It should work, and I'm all for it, except—how do you end up with ROOT files in the browser? They have to be explicitly copied in or something. Maybe the HTTP backend? They couldn't be very large...

The HTTP backend is what we're interested in for accessing ntuples that are on grid storage and we already have a service in LHCb which provides short lived EOS tokens as URL parameters to provide authenticiation. I think it's potentially interesting for quick browsing without needing to go the the hassle of setting up anything locally or have dedicated compute resources assigned to a jupyterhub instance.

**Comment by ioanaif** [comment]
@agoose77 yes, I'll take on the uproot side!

**Comment by chrisburr** [comment]
@ioanaif did you get this working or should this be re-opened? As far as I can tell there are still more issues to resolve for pyodide.

**Comment by ioanaif** [comment]
Hi! 

The issue reported here `RuntimeError: can't start new thread` was fixed in #868. 

However, `uproot` does not work at the moment in jupyterlite because we cannot import 'HTTPSConnection' which seems to be caused by the fact that `hhtp-client` is not available for pyodide yet. 

`ImportError: cannot import name 'HTTPSConnection' from 'http.client' (/lib/python311.zip/http/client.py)`
